Initialize cachedir

gperciva opened the pull request:
This is useful for the GUI in https://github.com/Tarsnap/tarsnap-gui/issues/45

cperciva commented on tar/chunks/chunks_stats.c:
```diff
@@ -414,3 +414,45 @@ chunks_stats_free(CHUNKS_S * C)
+int
+chunks_initialize(const char * cachepath)
+{
+	struct chunks_stats_internal * C;
```
We're not returning a cookie to be passed back to us later, so we don't need this.

cperciva commented on tar/chunks/chunks_stats.c:
```diff
@@ -414,3 +414,45 @@ chunks_stats_free(CHUNKS_S * C)
+	if ((C->cachepath = strdup(cachepath)) == NULL)
+		goto err1;
+
+	/* Read directory (a useful way to initialize C). */
```
Unnecessarily apologetic.

cperciva commented on tar/chunks/chunks_stats.c:
```diff
@@ -414,3 +414,45 @@ chunks_stats_free(CHUNKS_S * C)
+	/* Create a copy of the path. */
+	if ((C->cachepath = strdup(cachepath)) == NULL)
+		goto err1;
+
```
Check that the ${cachepath}/directory doesn't exist.  We should throw an error if it does, because the user is probably making a mistake...

cperciva commented on tar/chunks/chunks_stats.c:
```diff
@@ -414,3 +414,45 @@ chunks_stats_free(CHUNKS_S * C)
+	C->dir = dir;
+
+	/* Write directory file to filesystem. */
+	if (chunks_fsck_end(C))
```
Eh?  Why are we detouring through _fsck_ code here?

gperciva replied:
1. because that function contains `chunks_directory_write()`, and frees the `CHUNK_S` cookie.  Despite the name, it doesn't look particularly "fsck-related" to me.

2. because I can't resist calling a function whose description contains "... , and close the fscking cookie".  :)

cperciva commented on tar/chunks/chunks_stats.c:
```diff
@@ -414,3 +414,45 @@ chunks_stats_free(CHUNKS_S * C)
+	/* Write directory file to filesystem. */
+	if (chunks_fsck_end(C))
+		goto err0;
+	if (chunks_transaction_commit(cachepath))
```
Not necessary here.  The point of the `_transaction_commit` bit is to have state kept consistent with the server even in the event of an ill-timed crash; but we're not committing anything on the server here.

gperciva replied:
if I use `chunks_fsck_end` (which might not be suitable), then I'm left with a `directory.tmp` file.  `chunks_transaction_commit` renames that to `directory`.  (in a crash-proof way)

gperciva commented:
Updated PR.  Since the chunk, multitape, and glue layers changed so much, I merged the new code with the old commits, instead of adding them as new commits.  Future changes will be added as new commits (unless the whole architecture changes).

cperciva commented on tar/chunks/chunks_stats.c:
```diff
@@ -414,3 +414,50 @@ chunks_stats_free(CHUNKS_S * C)
+	/* Bail if ${chunkpath}/directory already exists. */
+	if (asprintf(&directory_filename, "%s/directory", cachepath) == -1)
+		goto err0;
+	if (access(directory_filename, F_OK) == 0)
```
Use `stat`, check for ENOENT.

cperciva commented on tar/chunks/chunks_stats.c:
```diff
@@ -414,3 +414,50 @@ chunks_stats_free(CHUNKS_S * C)
+chunks_initialize(const char * cachepath)
+{
+	char * directory_filename;
+	RWHASHTAB * ht;
```
It's `HT` everywhere else in this file.  Why different here?

cperciva commented on tar/chunks/chunks_stats.c:
```diff
@@ -414,3 +414,50 @@ chunks_stats_free(CHUNKS_S * C)
+	struct chunkstats stats_extra;
+
+	/* Bail if ${chunkpath}/directory already exists. */
+	if (asprintf(&directory_filename, "%s/directory", cachepath) == -1)
```
This might make sense as a `chunks_directory_exists` function?  That's the only file where the name "/directory" appears.

cperciva commented on tar/multitape/multitape.h:
```diff
@@ -205,6 +205,14 @@ int statstape_close(TAPE_S *);
+ * works.  This requires the "directory" file, but no other files.  Return 0
+ * on success, -1 on error, and 1 if the cachedir is already initialized.
+ */
+int initialize_cachedir(uint64_t, const char *);
```
This seems like an odd name as far as multitape function goes.  Something like `statstape_initialize` might make more sense?

cperciva commented on tar/multitape/multitape_fsck.c:
```diff
@@ -573,3 +573,26 @@ fscktape(uint64_t machinenum, const char * cachedir, int prune, int whichkey)
+initialize_cachedir(uint64_t machinenum, const char * cachedir)
+{
+	int rc;
+
```
`(void) machinenum; /* UNUSED */`

cperciva commented on tar/glue/tape.c:
```diff
@@ -232,6 +232,33 @@ tarsnap_mode_fsck(struct bsdtar *bsdtar, int prune, int whichkey)
+		break;
+	case 1:
+		bsdtar_warnc(bsdtar, 0, "Cache directory already initialized");
+		bsdtar->return_value = 1;
```
Not necessary here since err0 sets it...

gperciva commented:
Updated PR, with one commit per comment.  It needs history rewriting before a merge, of course.

I'm not totally sold on the name `statstape_initialize`, but I admit that the previous name wasn't appropriate either.  I've changed it as you requested.

cperciva commented on tar/chunks/chunks_directory.c:
```diff
@@ -362,6 +362,46 @@ chunks_directory_write(const char * cachepath, RWHASHTAB * HT,
+	return (1);
+
+fileexists:
+	/* File exists. */
```
memory leak?

cperciva commented on tar/chunks/chunks_directory.c:
```diff
@@ -362,6 +362,46 @@ chunks_directory_write(const char * cachepath, RWHASHTAB * HT,
 
 /**
+ * chunks_directory_exists(cachepath):
+ * Return 0 if the /directory file exists within ${cachepath}, 1 if it does
```
Seems like it would make more sense to return 1 for `_exists` if the file exists?

gperciva commented:
PR updated.  I think it might need another round of review after I've rewritten all this history.

Note that I did not change the returned values from `chunks_initialize`, even though `chunks_directory_exists` changed.  (I did update the `switch` statement inside `chunks_directory_exists`, though)

cperciva commented on tar/chunks/chunks_directory.c:
```diff
@@ -362,6 +362,48 @@ chunks_directory_write(const char * cachepath, RWHASHTAB * HT,
+		if (errno != ENOENT)
+			goto err1;
+	}
+
```
Given the three-way replication of freeing `directory_filename`, I'm tempted to break the usual style rule of "one return() statement per meaning" here.  `int rc;` which gets set to 0/1/-1, then `/* Return result code. */` `return (rc);`...

cperciva replied:
We should probably also `warnp("stat(%s)", directory_filename)` in the not-ENOENT error case, btw.

gperciva commented:
Updated PR.  Includes #151 , which will probably be merged before this one, so I can remove it from this PR when I clean up the history.

cperciva commented:
Looks good.  Please clean up the history, then I'll merge it.

gperciva commented:
Updated PR with clean history, and added a mention to the NEWS file.
